README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
'upwaver'
---------

'upwaver' is a wrapper around the UpWave API that allows users to access the UpWave API from within R.

Installation
------------

You can install 'upwaver' from `CRAN` or `github` with:

``` r
install.packages("upwaver")

# install.packages("devtools")
devtools::install_github("ims-fhs/upwaver")
```

Examples
--------

You can access the following UpWave informations and more with 'upwaver':

-   List of Boards: The list of your UpWave Boards with `list_boards()`
-   Board details: Background Information about a specific board with `board_details()`
-   List of Cards: Detailed Information about all Cards of a specific board with `list_cards()`

``` r
# upwaver::list_boards("your_organisation", "your_token")
# upwaver::board_details("your_organisation", board_id, "your_token")
# upwaver::list_cards("your_organisation", board_id, "your_token")
```

For more detailed Information and further functionalities of the 'upwaver' package, check out the vignette.


NEWS.md
# upwaver 1.2.1

* Minor changes in tests to reflect a minor change in the 'Upwave' API endpoint. The API of the package did not change.

# upwaver 1.2.0

* Minor changes to reflect new 'Upwave' API endpoint. The API of the package did not change.

# upwaver 1.1.0

* New function storypoints_and_status_released(): storypoints for all cards with status 'released'.
* Bugfix: Catch cases where there are more than 100 cards on a board since storypoint calculation might be wrong in those cases.
* Change one testcase to reflect 'Upwave' API-Changes.

# upwaver 1.0.1

* Change of API-Token in vignette.
* Renaming (upwaver -> 'upwaver' / UpWave -> 'Upwave') in vignette and documentation
* Bugfix in tests.

# upwaver 1.0.0

* Added a `NEWS.md` file to track changes to the package.
* Basic functionality as described in readme.md and vignette.


tests/testthat/test-upwaver.R
context("low level functions")

test_that("request() works", {
  response <- request("https://ims-fhs.upwave.io/api/boards/",
                               "807a0c3451c04602b4dbfdc0338a65f7")

  expect_that(response, is_a("response"))
  expect_that(response$status_code, equals(200))
})

test_that("content_as_list() works", {
  result <- content_as_list(request(
    "https://api.upwave.io/workspaces/4597/boards/",
    "807a0c3451c04602b4dbfdc0338a65f7"))

  expect_that(result, is_a("list"))
  expect_that(length(result), equals(4))
})

context("Board list and details")

test_that("list_boards() works", {
  list_boards <- list_boards("4597", "807a0c3451c04602b4dbfdc0338a65f7")
  expect_that(list_boards, is_a("list"))
  expect_that(length(list_boards), equals(4))

  list_boards <- list_boards("99", "807a0c3451c04602b4dbfdc0338a65f7")
  expect_that(list_boards, is_a("list"))
  expect_that(length(list_boards), equals(2))
  expect_that(list_boards$detail, equals("Not found."))

  board_details <- board_details("99", 1234, "807a0c3451c04602b4dbfdc0338a65f7")
  expect_that(board_details, is_a("list"))
  expect_that(length(board_details), equals(2))
  expect_that(board_details$detail, equals("Not found."))
})

test_that("board_details() works", {

  id <- 14351
  board_details <- board_details("4597", id, "807a0c3451c04602b4dbfdc0338a65f7")
  expect_that(board_details, is_a("list"))
  expect_gt(length(board_details), 1)
  expect_that(board_details$id, equals(id))

})


context("Card list and details")
test_that("list_cards() works", {
  list_cards <- list_cards("4597", 14351, "807a0c3451c04602b4dbfdc0338a65f7")
  expect_that(list_cards, is_a("list"))
  expect_that(length(list_cards), equals(4))
})

context("Write boards and cards")
test_that("create_card_from_excel() works", {
  # token <- "807a0c3451c04602b4dbfdc0338a65f7"
  # res <- create_card_from_excel(excel_file = "inst/extdata/helper_user_stories.xlsx",
  #                               sheet = 2, line_number = 1, organisation = "ims-fhs", board_id = 14351,
  #                               board_name = "upwaver_tests", token = token)
  #
  # expect_that(res$status_code, equals(200))
  # expect_error(
  #   create_card_from_excel(organisation = "ims-fhs", board_id = 14351, board_name = "upwaver_tests",
  #            token = token, excel_file = "helper_user_stories.xlsx", sheet = 2, line_number = 1))
})

context("Storypoints")
test_that("storypoints() works", {
  storypoints <- storypoints_and_status("4597", 14351, "807a0c3451c04602b4dbfdc0338a65f7")
  expect_that(storypoints, is_a("list"))
  expect_that(storypoints$sum_storypoints, is_a("numeric"))
})

test_that("aufwand() works", {
  expect_equal(aufwand("Aufwand: 10 AT"), 10)
  expect_equal(aufwand("Aufwand: 10 SP"), 10)
  expect_equal(aufwand("blabla"), 0)
  expect_error(aufwand(3))
})


R/burndown.R
#' Calculate storypoints ans status of the Cards of a specific Board
#'
#' Calculates storypoints of released cards on a specific board
#'
#' @param organisation A character. The name of the organisation as specified in the URL:
#' https://<organisation>.upwave.io
#' @param board_id An integer. The number of the board you want to inspect as specified in
#' the URL: https://<organisation>.upwave.io/board/<board_id>/view/
#' @param token A character. The API token to access boards on upwave.io. The token can be
#' obtained by visiting your account settings. Click your profile image, select "Settings"
#' and find your API-Key in the "Account" tab.
#'
#' @return A list containing a data.frame with all storypoints and the status per card and
#' the numeric sum_storypoints.
#' @export
#'
#' @examples upwaver::storypoints_and_status_released(
#' "4597", 14351, "807a0c3451c04602b4dbfdc0338a65f7")
storypoints_and_status_released <- function(organisation, board_id, token) {
  cards <- upwaver::list_cards(organisation, board_id, token)$results
  if (length(cards) >= 100) {stop("More than 100 cards on board, calculation of storypoints may be wrong.")}

  df <- data.frame(title = character(length(cards)),
                   status = numeric(length(cards)),
                   storypoints = character(length(cards)),
                   stringsAsFactors = F)
  for (i in 1:length(cards)) {
    # state 0: Not specified, 1: Not started, 2: In progress, 3: Completed
    df[i, 1] <- cards[[i]]$title
    df[i, 2] <- cards[[i]]$state
    if (cards[[i]]$state == 3) {
      description <- cards[[i]]$description
      df[i, 3] <- aufwand(description)
    } else {
      df[i, 3] <- "Active" # active tasks
    }
  }
  return(list(storypoints_status = df,
              sum_storypoints = sum(as.numeric(df$storypoints[df$storypoints != "Active"]))))
}


#' Calculate storypoints ans status of the Cards of a specific Board
#'
#' Calculates storypoints (of active Cards, which are Cards that haven't been completed
#' yet) and status in the backlog of an 'Upwave' board.
#' Storypoints have to be specified in the description (directly beneath the title) of
#' each card with a string of the form: "Aufwand: X SP" or "Aufwand: X AT". This is not
#' part of the 'Upwave' API but an additional functionality.
#'
#' @param organisation A character. The name of the organisation as specified in the URL:
#' https://<organisation>.upwave.io
#' @param board_id An integer. The number of the board you want to inspect as specified in
#' the URL: https://<organisation>.upwave.io/board/<board_id>/view/
#' @param token A character. The API token to access boards on upwave.io. The token can be
#' obtained by visiting your account settings. Click your profile image, select "Settings"
#' and find your API-Key in the "Account" tab.
#'
#' @return A list containing a data.frame with all storypoints and the status per card and
#' the numeric sum_storypoints.
#' @export
#'
#' @examples
#' upwaver::storypoints_and_status("4597", 14351, "807a0c3451c04602b4dbfdc0338a65f7")
storypoints_and_status <- function(organisation, board_id, token) {
  cards <- upwaver::list_cards(organisation, board_id, token)$results
  cards <- lapply(cards, function(x){x$description <- gsub("<.*?>", "", x$description); x})
  cards <- lapply(cards, function(x){x$description <- gsub("\\&\\#\\d{5};", "", x$description); x})
  if (length(cards) >= 100) {stop("More than 100 cards on board, calculation of storypoints may be wrong.")}

  df <- data.frame(title = character(length(cards)),
                   status = numeric(length(cards)),
                   storypoints = character(length(cards)),
                   stringsAsFactors = F)
  for (i in 1:length(cards)) {
    # state 0: Not specified, 1: Not started, 2: In progress, 3: Completed
    df[i, 1] <- cards[[i]]$title
    df[i, 2] <- cards[[i]]$state
    if (cards[[i]]$state != 3) {
      description <- cards[[i]]$description
      df[i, 3] <- aufwand(description)
    } else {
      df[i, 3] <- "Completed" # completed tasks
    }
  }
  return(list(storypoints_status = df,
              sum_storypoints = sum(as.numeric(df$storypoints[df$storypoints != "Completed"]))))
}


#' Calculates the effort of a given card (given by its description)
#'
#' aufwand() parses the description of a card for tags like "Aufwand: 10 AT" or
#' "Aufwand: 10 SP".
#'
#' @param description A string extracted from an 'Upwave' card
#'
#' @return An integer
aufwand <- function(description) {
  assertthat::is.string(description)

  aufwand <- strsplit(description, split = "Aufwand: ")[[1]][2]
  aufwand <- strsplit(aufwand, split = "[ SP| AT]")[[1]][1]
  if(is.na(aufwand)) {aufwand <- 0}
  assertthat::assert_that(assertthat::noNA(aufwand))

  return(as.numeric(aufwand))
}


R/board2excel.R
#' Save all information from a board to Excel
#'
#' @param organisation A character
#' @param board_id A character
#' @param token A character
#' @param filename A character
#' @param path A character
#'
#' @export
#'
#' @examples
#' \dontrun{
#' organisation <- "4597"
#' board_id <- 14351
#' token_sqc <- "807a0c3451c04602b4dbfdc0338a65f7"
#' upwaver::board2excel(organisation, board_id, token_sqc, "test.xlsx", "C:/Temp/")
#' }
board2excel <- function(organisation, board_id, token, filename, path) {
  upwaver <- list_cards(organisation, board_id, token)
  ff <- upwaver$results

  n <- length(ff)
  all_cards <- data.frame(id = character(n),
                          title = character(n),
                          text = character(n),
                          tasks = character(n), stringsAsFactors = FALSE)

  for (i in 1:n) {
    card <- ff[[i]]
    all_cards$id[i] <- card$id
    all_cards$title[i] <- card$title

    # description
    textline <- card$description
    textline <- gsub("<br/>", "\n", textline)
    splits <- gregexpr("<.*?>", textline)[[1]]
    len <- attributes(splits)$match.length
    textline <- substr(textline, splits[2] + len[2], nchar(textline))
    all_cards$text[i] <- gsub("<.*?>", "", textline)

    # tasks:
    url <- paste0("https://", organisation, ".upwave.io/api/tasklistitems/?card=", card$id)
    upwaver_tasks <- content_as_list(request(url, token))$results
    task_list <- c()
    if (length(upwaver_tasks) > 0) {
      for (k in 1:length(upwaver_tasks)) {
        task_list <- c(task_list, upwaver_tasks[[k]]$description)
      }
      all_cards$tasks[i] <- paste(task_list, collapse="\n")
    } else {
      all_cards$tasks[i] <- ""
    }
  }
  # save Excel
  if (file.exists(paste0(path, filename))) {
    message(paste0("Old file ", paste0(path, filename), " overwritten"))
  }
  xlsx::write.xlsx(all_cards, paste0(path, filename), row.names = FALSE)
}





R/create_card.R
#' Create a new card on a specified 'Upwave' board
#'
#' create_card_from_excel() adds a new card in the most-left column of the given board.
#' The Information of the card are taken from an excel file.
#'
#' @param excel_file A character. The path to an excel-file providing informations for user-stories;
#' expected columns: US-ID, name, description
#' @param sheet An integer or a character. The sheet of the excel-file providing informations for user-stories;
#' expected columns: US-ID, name, description
#' @param line_number An integer. The line number in the spreadsheet starting from 1
#' @param organisation A character. The name of the organisation as specified in the URL:
#' https://<organisation>.upwave.io
#' @param board_id An integer. The board number, the card has to be created in
#' @param board_name A character. The board name, the card has to be created in
#' @param token A character. The API token to access boards on upwave.io. The token can be
#' obtained by visiting your account settings. Click your profile image, select "Settings"
#' and find your API-Key in the "Account" tab.
#'
#' @export
#'
#' @examples
#' \dontrun{create_card_from_excel(excel_file = "path_to_your_excel_file", sheet = (sheetno. or sheetname),
#' line_number = linenr, organisation = "your_organisation", board_id = your_board_id,
#' board_name = "your_board_name", "token = your_api_token")}
create_card_from_excel <- function(excel_file, sheet, line_number, organisation,
                                   board_id, board_name, token) {
  name_by_board_id <- board_details(organisation, board_id, token)$title
  if (length(name_by_board_id) == 0) {
    stop(paste0("Board ", board_name, " does not exist"))
  }
  assertthat::assert_that(name_by_board_id == board_name)

  url <- paste0("https://", organisation, ".upwave.io/api/cards/")

  df <- XLConnect::readWorksheetFromFile(excel_file,
                                         sheet= sheet,
                                         startRow = 1,
                                         endCol = 100)

  data <- df[line_number,]

  json <- list(description = paste0("<h1>US",data$US.ID," - ",
                                    data$name, "</h1><p>",
                                    data$description, "</p>"),
               board = board_id)

  httr::POST(url, httr::content_type("application/json"),
             httr::add_headers(Authorization = paste0("Token ", token)),
             body = json, encode = "json")
}


R/list_cards_pretty.R
#' #' Save all cards from a board to a data.frame
#' #'
#' #' @param organisation A character
#' #' @param board_id A character
#' #' @param token A character
#' #'
#' #' @export
#' #'
#' #' @examples
#' #' organisation <- "4597"
#' #' board_id <- 14351
#' #' token_sqc <- "807a0c3451c04602b4dbfdc0338a65f7"
#' #' upwaver::list_cards_pretty(organisation, board_id, token_sqc)
#' list_cards_pretty <- function(organisation, board_id, token) {
#'   stop("Does not work with new Upwaver API.")
#'   # upwaver <- list_cards(organisation, board_id, token)
#'   # ff <- upwaver$results
#'   #
#'   # n <- length(ff)
#'   # all_cards <- data.frame(id = character(n),
#'   #                         title = character(n),
#'   #                         text = character(n),
#'   #                         tasks = character(n), stringsAsFactors = FALSE)
#'   #
#'   # for (i in 1:n) {
#'   #   card <- ff[[i]]
#'   #   all_cards$id[i] <- card$id
#'   #   all_cards$title[i] <- card$title
#'   #
#'   #   # description
#'   #   textline <- card$description
#'   #   splits <- gregexpr("<.*?>", textline)[[1]]
#'   #   len <- attributes(splits)$match.length
#'   #   textline <- substr(textline, splits[2] + len[2], nchar(textline))
#'   #   all_cards$text[i] <- gsub("<.*?>", "", textline)
#'   #
#'   #   # tasks:
#'   Most probably this line needs to be changed with the new API endpoints. SCN
#'   #   url <- paste0("https://", organisation, ".upwave.io/api/tasklistitems/?card=", card$id)
#'   #   upwaver_tasks <- content_as_list(request(url, token))$results
#'   #   task_list <- c()
#'   #   if (length(upwaver_tasks) > 0) {
#'   #     for (k in 1:length(upwaver_tasks)) {
#'   #       task_list <- c(task_list, upwaver_tasks[[k]]$description)
#'   #     }
#'   #     all_cards$tasks[i] <- paste(task_list, collapse="<br>")
#'   #   } else {
#'   #     all_cards$tasks[i] <- ""
#'   #   }
#'   # }
#'   # all_cards
#' }
#'
#'
#'


R/upwaver.R
#' Information about your organisation's boards
#'
#' list_boards() is a wrapper around "List Boards" as specified in the 'Upwave' API
#' https://www.upwave.io/api/. It returns information for your organisation's
#' boards as id's, title's, time of creation and the users who created the boards.
#'
#' @param organisation A character. The name of the organisation as specified in the URL:
#' https://<organisation>.upwave.io
#' @param token A character. The API token to access boards on upwave.io. The token can be
#' obtained by visiting your account settings. Click your profile image, select "Settings"
#' and find your API-Key in the "Account" tab.
#'
#' @return A list
#' @export
#'
#' @examples
#' list_boards("4597", "807a0c3451c04602b4dbfdc0338a65f7")
list_boards <- function(organisation, token) {
  url <- paste0("https://api.upwave.io/workspaces/", organisation, "/boards/")
  return(content_as_list(request(url, token)))
}


#' Information about a specific board
#'
#' board_details() is a wrapper around "board_details" as specified in the 'Upwave' API
#' https://www.upwave.io/api/. It returns detailed information about a specific board such
#' as id, title, purpose, background image, creation time, creator, columns, rows and colors.
#'
#' @param organisation A character. The name of the organisation as specified in the URL:
#' https://<organisation>.upwave.io
#' @param board_id An integer. The number of the board you want to inspect as specified in
#' the URL: https://<organisation>.upwave.io/board/<board_id>/view/
#' @param token A character. The API token to access boards on upwave.io. The token can be
#' obtained by visiting your account settings. Click your profile image, select "Settings"
#' and find your API-Key in the "Account" tab.
#'
#' @return A list
#' @export
#'
#' @examples
#' board_details("4597", 14351, "807a0c3451c04602b4dbfdc0338a65f7")
board_details <- function(organisation, board_id, token) {
  url <- paste0("https://api.upwave.io/workspaces/", organisation, "/boards/", board_id, "/")
  return(content_as_list(request(url, token)))
}


#' Information about the cards on a board
#'
#' list_cards() is wrapper around "List Cards" as specified in the 'Upwave' API
#' https://www.upwave.io/api/. It returns detailed Information about each card of a specific
#' board as card id, title, board id, description, due_dt, state, color, assigned, ...
#'
#' @param organisation A character. The name of the organisation as specified in the URL:
#' https://<organisation>.upwave.io
#' @param board_id An integer. The number of the board you want to inspect as specified in
#' the URL: https://<organisation>.upwave.io/board/<board_id>/view/
#' @param token A character. The API token to access boards on upwave.io. The token can be
#' obtained by visiting your account settings. Click your profile image, select "Settings"
#' and find your API-Key in the "Account" tab.
#'
#' @return A list
#' @export
#'
#' @examples
#' list_cards("4597", 14351, "807a0c3451c04602b4dbfdc0338a65f7")
list_cards <- function(organisation, board_id, token) {
  url <- paste0("https://api.upwave.io/workspaces/", organisation, "/cards/?board=", board_id)
  return(content_as_list(request(url, token)))
}


#' Create a correct request for the 'Upwave' API
#'
#' request() is an internal function that invokes a GET request with a token according to
#' https://www.upwave.io/api/
#'
#' @param url an URL
#' @param token an authentication token
#'
#' @return request
request <- function(url, token) {
  return(httr::GET(url, httr::add_headers(Authorization = paste0("Token ", token))))
}

#' Transform the answer of a 'Upwave' API request to a list
#'
#' content_as_list() is an internal function that extracts content from a request and
#' coerces it to a list format
#'
#' @param response An object of class "response"; response of a GET request
#'
#' @return A list
content_as_list <- function(response) {
  json <- httr::content(response, as = "text")
  return(rjson::fromJSON(json))
}
